### DISCUSSION TITLE
Low power mode

### DISCUSSION THREAD

**Discussion by laurensvalk** [pybricks/support#623] [Ideas]
We occasionally get requests to store some runtime data on the City Hub, Technic Hub, and Move Hub.

Since we are not currently enabling flash access for safety reasons, an alternative would be to allocate some static memory and have a getter/setter to read/write some bytes from/to it from a user program. This would be easy enough to do.

This would allow keeping some data between programs. Use cases:
- Store state of motors so you don't have to initialize or re-align them every time
- Maintain data or counters that shouldn't be reset every time

For example, a program counter could look like this:

```python
from pybricks.tools import data_storage

# Read the currently stored data
data = data_storage()

# Get counter data
count = data[0]
print(count)

# Increment the counter store the result.
data_storage(bytes([count + 1]))

```
A one-off motor initialization could look like this:

```python
from pybricks.tools import data_storage

if not data_storage():
    print("This is the first time, so initialize the motors once now.")
    data_storage(b"1")
```

But it would be even more useful if the data could be preserved for longer, while the hub is "off".

In addition to the above, this could then be used to store things like user settings.

A use case could look like this:

```python
from pybricks.hub import CityHub
from pybricks.tools import data_storage

hub = CityHub()

print(data_storage())
data_storage(b"Store this!")

hub.system.sleep()
```

`hub.system.sleep()` would do everything `system.shutdown()` does, but it keeps the `stm32` running, so preserving RAM.

With all external resources off, the stm32 itself might not even need to enter a low power mode to make this useable, but it might be a nice touch.

The key thing that makes this simple, is that we're not trying to keep the MicroPython program running. So it would work and feel just like a normal shutdown, it's just that `data_storage()` remains populated.

**Comment by falk12** [comment]
Hi,

thanks for  starting this topic.
Will the data be available  when batteries had been removed?

**Reply by laurensvalk** [reply]
That remains to be seen. It depends on how much power it still uses in that mode. Maybe it gives you a few minutes to replace the battery. That was the case for the good old RCX too.

But even if you had to re-initialize just once after swapping the batteries, that's still better than doing it every time.

**Reply by dlech** [reply]
Without a separate power supply, like a coin-cell battery, which the hubs don't have, it is physically impossible to prevent the MCU from losing power when the battery is removed. The RCX must have been specifically designed to do that (huge capacitors?). So there is no way for data to persist across battery changes on the new hubs. The majority of the power draw on the hubs comes from passive devices, not the MCU so even if the MCU was in a low power mode, it would only make a few mA difference in current consumption out of 10s of mA used by the passive components.

**Comment by ZPhilo** [comment]
Yes, RCX had big input caps to allow battery change without loosing memory contents (https://brickshelf.com/gallery/Philo/RCX-comp/rcx1-top.jpg, bottom right). BTW, these were an efficient way to kill your RCX: if you accidentally shorten power terminals while changing the battery, the caps discharged through input fuse, instantly blowing it!
But if you manage to put the processor in low power mode, there's a good probability to be able to change batteries without loosing memory as the Pup hubs does have a large input cap too.
(Technic hub https://i.stack.imgur.com/s3wAS.jpg, City hub https://farm2.staticflickr.com/1732/27768363217_61aca1d757_o.png, Boost hub https://i.stack.imgur.com/i9YZp.jpg)

**Reply by dlech** [reply]
> instantly blowing it!

Fun times. 😄 

> Pup hubs does have a large input cap too.

From my experience, this is enough to power the Technic hub for about 1, maybe 2 seconds. When we were first developing the firmware, it was common to have to pull the batteries if the firmware crashed. Since it is easy to slide the batteries in and out of this hub I would do it rather quickly. If I did it too quick though, the power would not be lost and the hub would not boot.

**Comment by ZPhilo** [comment]
> From my experience, this is enough to power the Technic hub for about 1, maybe 2 seconds.

Yes, but this is somehow expected if the hub was fully active. But I definitely have some concerns if care was not taken to create an electronics able to have a real low power mode. And that was clearly not the target of the PUP hubs!

**Comment by vascolp** [comment]
It looks cool, but I dislike not being able to remove batteries for long periods.
Instead of low power mode, couldn’t you set up a flash area like the name but with say 128bytes, plus a timestamp or StopWatch, to avoid repeated write attempts?
Something that would allow only one write per second or 30s or whatever?
It would be "slow writing" but battery removable proof…
